### PR TITLE
[RLlib] PyFlyt QuadX WayPoints

### LINKED ISSUE #44738: [RLlib] Add PyFlyte Example Using Gymnasium and Ray Tune
### Description

Add a PPO RLlib + Ray Tune that trains an agent on a [PyFlyt environment](https://github.com/jjshoots/PyFlyt).

Additionally, this example will showcase how to register an environment in gymnasium while leveraging their wrappers to adjust the reward function to maximize learning potential.

### Use case

Showcase how to:
- register a gymnasium environment for use with Ray Tune and RLlib.
- wrap and adjust environment functions using Gymnasium wrappers as a potential solution instead of adjusting hyperparameters from the start.
- train an agent in a difficult observation space.

### PR DESCRIPTION
**Pull request by peterghaddad**
## Why are these changes needed?

An example to leverage the PyFlyt environment with RLlib enabling users to train drones in the PyBullet physics engine.

## Related issue number

Closes #44738 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by peterghaddad** [comment]
Hi @ArturNiederfahrenhorst! Is it possible to get your eyes on this PR? Thanks in advance!

**Comment by anyscalesam** [comment]
@sven1977 @simonsays1980 can you take a look please?

File: rllib/examples/envs/classes/pyflyt_quadx_waypoints_env.py
Comment (simonsays1980) [inline rllib/examples/envs/classes/pyflyt_quadx_waypoints_env.py:23]: Very nice!

Comment (simonsays1980) [reply rllib/examples/envs/classes/pyflyt_quadx_waypoints_env.py:23]: @peterghaddad Can we move this code to the main file below and keep a single-file script?

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:16]: We are moving to a new API stack that supports our `RLModule API`, `Learner API`, and `EnvRunner API`, mainly for better customization and performance. Therein, we support mainly PyTorch (for PPO and IMPALA also `tf2` but no `tf` anymore). 

Could we adjust this arg towards this?

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:53]: Small nit: Can we format this in a way such that each method is in a new line? For an example take a look into the other example files. That makes it easier for users to grab an overview.

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:62]: Can we support here solely the new stack? The reason is that we want new users to jump directly onto the new stack. Here this would mean it needs in addition the following calls to the `config`:

```python
.api_stack(
            enable_rl_module_and_learner=True,
            enable_env_runner_and_connector_v2=True,
)
.rl_module(
            model_config_dict={
                "use_lstm": True,
                "lstm_cell_size": 32,
                "lstm_use_prev_action": args.use_prev_action,
                "lstm_use_prev_reward": args.use_prev_reward,
            }
)
``` 
The `model` argument needs then to vanish from `training()`.

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:73]: I know, its already enough :) WIth the new stack come also new metrics and our `MetricsLogger API` that makes logging custom metrics easier. The names have changed with this a bit, .e.g `"episode_return_mean"`.

**Review by simonsays1980 — CHANGES_REQUESTED** [review]
Awesome PR - the example is very nice! CHanges requested concern mainly our move from the old to the new stack. @peterghaddad if you need help somewhere or want me to adapt the code, ping me please.

We appreciate your work @peterghaddad ! This is really a nice example I have not thought about before. Hope to see more from you in the future!

File: rllib/examples/quadx_waypoints.py
Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:82]: Hi @simonsays1980 of course!

I do see some errors occurring when converting over to the new stack. Most likely the observation space, but it might help others down the line for seeing this. Let me know if you have any thoughts, i'll start digging into it!
```
ValueError: The two structures don't have the same nested structure.

First structure: type=dict str={'actor': {'h': array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]],
      dtype=float32), 'c': array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]],
      dtype=float32)}, 'critic': {'h': array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]],
      dtype=float32), 'c': array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]],
      dtype=float32)}}

Second structure: type=dict str={'h': array([[[-0.05538796,  0.13966578,  0.01491474, ...,  0.02647178,
          0.048148  , -0.07149553]],

       [[ 0.10487001, -0.12613896,  0.1936225 , ..., -0.04218567,
          0.02297537, -0.04598944]],

       [[-0.0181545 ,  0.20065415, -0.03715083, ...,  0.02511652,
          0.0963103 , -0.00859845]],

       ...,

       [[-0.10791852,  0.21027309,  0.04330226, ...,  0.06297404,
         -0.1705255 ,  0.20365342]],

       [[ 0.08050264,  0.1678762 ,  0.14637923, ...,  0.08334337,
         -0.09527908,  0.21276303]],

       [[ 0.08050264,  0.1678762 ,  0.14637923, ...,  0.08334337,
         -0.09527908,  0.21276303]]], dtype=float32), 'c': array([[[-0.10788506,  0.2418374 ,  0.03368664, ...,  0.05303482,
          0.12481229, -0.16512036]],

       [[ 0.2415701 , -0.25722432,  0.3673103 , ..., -0.07641751,
          0.05405141, -0.0850858 ]],

       [[-0.027573  ,  0.2839686 , -0.11287448, ...,  0.05270897,
          0.36143133, -0.02489884]],

       ...,

       [[-0.26991454,  0.49751803,  0.07710288, ...,  0.1402412 ,
         -0.41773236,  0.5066306 ]],

       [[ 0.20296642,  0.47725105,  0.25387886, ...,  0.16696252,
         -0.29451394,  0.40619874]],

       [[ 0.20296642,  0.47725105,  0.25387886, ...,  0.16696252,
         -0.29451394,  0.40619874]]], dtype=float32)}
```

Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:82]: Looks like it has to do with the depth of the array between the state_out and the look_back_state.
I am looking to see the differences between the new API and older api. Unsure if this is fixable through an observation wrapper or if it's better to create a new issue. @simonsays1980 Let me know your thoughts!

Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:79]: I ruled out it was because LSTM is enabled. Seems to be some problems using it with the new API. I can create another issue for it to be addressed within the new API. Think we're good here for now.

**Comment by peterghaddad** [comment]
@simonsays1980 addressed feedback! Let me know if there is anymore feedback!

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [reply rllib/examples/quadx_waypoints.py:83]: @peterghaddad Thanks for digging in. I can replicate this error. Yes this has to do with the LSTM and the connector. I have a guess where this is coming from. We introduced `inference-only` `RLModules` and these do not have a critic. We have to dig into this deeper.

**Comment by simonsays1980** [comment]
> @simonsays1980 addressed feedback! Let me know if there is anymore feedback!

@peterghaddad Thanks for the comments! So, for the error occuring in the new stack we need to dig into this deeper. I have a guess, where this is rooted. I come back to you when I know more.

**Comment by peterghaddad** [comment]
> > @simonsays1980 addressed feedback! Let me know if there is anymore feedback!
> 
> @peterghaddad Thanks for the comments! So, for the error occuring in the new stack we need to dig into this deeper. I have a guess, where this is rooted. I come back to you when I know more.

@simonsays1980 Appreciate you diving in! From a quality and training perspective, I think this example is good to go. LSTM isn't necessary for us to move forward; however, i'm glad we are finding things that can be fixed!

**Comment by peterghaddad** [comment]
@simonsays1980 are we good to move forward here? Also, thanks for diving into the fix use_lstm

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:31]: Great! This is now single file

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:68]: Top! Clean and smooth :)

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [reply rllib/examples/quadx_waypoints.py:83]: @peterghaddad Alright, my guess was right. I fixed it and its in master. Can you merge the newest master pls and check, if the error vanishes?

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [inline rllib/examples/quadx_waypoints.py:8]: We have a standard parser in `rllib/utils/test_utils.py` and a run script which we use for examples. This is something I can also do in a cleanup after merge. Your PR, your choice. As soon as this script runs through we are ready to run premerge tests imo.

**Review by simonsays1980 — CHANGES_REQUESTED** [review]
LGTM. Needs to pull the newest master with a fix for LSTM. Needs a cleanup before or after merge to include RLlib test utils.

File: rllib/examples/quadx_waypoints.py
Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:83]: Error still exists unfortunately! Updated the version of Ray by running:
`pip install -U "ray[all] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-3.0.0.dev0-cp39-cp39-macosx_10_15_x86_64.whl"`

I think we can move forward without lstm, and I can continue to troubleshoot it on the side. It seems to be the same error we were seeing before.

File: rllib/examples/quadx_waypoints.py
Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:8]: Going to give this a shot! Do these examples run against tests when changes are made in future PRs?

**Comment by peterghaddad** [comment]
Included RLLib test utils!

File: rllib/examples/quadx_waypoints.py
Comment (simonsays1980) [reply rllib/examples/quadx_waypoints.py:90]: @peterghaddad I cannot see in your fork the new changes - its still behind. Can you pull again from upstream master? From what I can see there are missing a lot of changes - should be also significant faster after the pull - we put a lot into performance and I can see `pyflyt` is quite fast, too :)

Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:90]: Yes. Think we're in action now.

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome PR !

**Comment by peterghaddad** [comment]
@simonsays1980 thanks for approving! Still seeing the same error with lstm. Let's create another issue and address that there since the agent is converges without lstm. Once you reapprove, will you also be the one to merge it in? Thanks again for all the review!

File: rllib/examples/quadx_waypoints.py
Comment (sven1977) [inline rllib/examples/quadx_waypoints.py:1]: Awesome stuff! Thanks for this PR @peterghaddad !

Looks good to me. After merging this, we should add it to the examples docs (which we'll also have to clean up a bit, still).

**Review by sven1977 — APPROVED** [review]
LGTM!

File: rllib/examples/quadx_waypoints.py
Comment (sven1977) [inline rllib/examples/quadx_waypoints.py:116]:
```suggestion
    stop = {
        TRAINING_ITERATION: args.stop_iters,
        f"{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}": args.stop_reward,
    }
```

File: rllib/examples/quadx_waypoints.py
Comment (sven1977) [inline rllib/examples/quadx_waypoints.py:31]:
```suggestion
```

File: rllib/examples/quadx_waypoints.py
Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py:1]: @sven1977 thank you for reviewing and your feedback. I addressed your comments.

I think that's a great idea. I'll write up an issue #45586

File: rllib/examples/quadx_waypoints.py
Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py]: Do we plan on changing `TRAINING_ITERATION_TIMER` to `TRAINING_ITERATION`? I don't see this constant in master.

Comment (peterghaddad) [reply rllib/examples/quadx_waypoints.py]: I removed the reference to the TRAINING_ITERATION constant because it does not exist. I changed it to the direct string value i.e `"training_iteration" args.stop_iters,`

Happy to adjust!

Comment (simonsays1980) [reply rllib/examples/quadx_waypoints.py]: @peterghaddad `TRAINING_ITERATION_TIMER` is the constant referring to the timer for a training iteration, while `TRAINING_ITERATION` refers to the `int` training iteration. The former comes indeed from our metrics constants the latter comes (still) from `ray.air.constants`.

**Comment by sven1977** [comment]
All good to go. Sorry @peterghaddad for the many delays! Will auto-merge this once all tests pass ...

**Pull request merged**
